src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -1492,29 +1509,16 @@ public sealed override bool IsTupleCompatible(out int tupleCardinality)
                     do
                     {
                         levelsOfNesting++;
-                        typeToCheck = ((NamedTypeSymbol)typeToCheck).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[TupleTypeSymbol.RestPosition - 1].Type;
-                    }
-                    while (TypeSymbol.Equals(typeToCheck.OriginalDefinition, this.OriginalDefinition, TypeCompareKind.ConsiderEverything2) && !typeToCheck.IsDefinition);
-
-                    if (typeToCheck.IsTupleType)
-                    {
-                        var underlying = typeToCheck.TupleUnderlyingType;
-                        if (underlying.Arity == TupleTypeSymbol.RestPosition && !TypeSymbol.Equals(underlying.OriginalDefinition, this.OriginalDefinition, TypeCompareKind.ConsiderEverything2))
-                        {
-                            tupleCardinality = 0;
-                            return false;
-                        }
-
-                        tupleCardinality = (TupleTypeSymbol.RestPosition - 1) * levelsOfNesting + typeToCheck.TupleElementTypesWithAnnotations.Length;
-                        return true;
+                        typeToCheck = ((NamedTypeSymbol)typeToCheck).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[NamedTypeSymbol.RestPosition - 1].Type;
                     }
+                    while (TypeSymbol.Equals(typeToCheck.OriginalDefinition, this.OriginalDefinition, TypeCompareKind.ConsiderEverything) && !typeToCheck.IsDefinition);
 
                     arity = (typeToCheck as NamedTypeSymbol)?.Arity ?? 0;
 
-                    if (arity > 0 && arity < TupleTypeSymbol.RestPosition && typeToCheck.IsTupleCompatible(out tupleCardinality))
+                    if (arity > 0 && arity < NamedTypeSymbol.RestPosition && ((NamedTypeSymbol)typeToCheck).IsTupleTypeOfCardinality(out tupleCardinality))
                     {
-                        Debug.Assert(tupleCardinality < TupleTypeSymbol.RestPosition);
-                        tupleCardinality += (TupleTypeSymbol.RestPosition - 1) * levelsOfNesting;
+                        Debug.Assert(tupleCardinality < NamedTypeSymbol.RestPosition);
+                        tupleCardinality += (NamedTypeSymbol.RestPosition - 1) * levelsOfNesting;
```

Comment:
This assertion can fail if you construct a `Tuple'8` with a `Rest` field containing a `Tuple'8` from another assembly.  It can also fail if the `Rest` field contains an instance of `this` inside the declaration of `Tuple'8`, where is `IsDefinition` will be true.

Reply:
In that first scenario, I think we will not get into this `if` statement. We'll have exited the `while` loop above because the original definitions of the two `ValueTuple'8` don't match, so `arity` will be great than 8, and we won't enter the `if`.

For the second scenario, I don't think we can have the `Rest` field be a `ValueTuple'8`, because a struct type cannot contain itself.

---
In reply to: [356905207](https://github.com/dotnet/roslyn/pull/39370#discussion_r356905207) [](ancestors = 356905207)